src/transformStateWithClones.js

```diff
@@ -1,13 +1,31 @@
-'use strict';
-
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const stateHistory = [];
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(stateClone, obj.extraData);
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const delObj of obj.keysToRemove) {
+        delete stateClone[delObj];
+      }
+      stateHistory.push({ ...stateClone });
```

Comment:
[DONT REPEAT YOURSELF]: Same as the previous comment, avoid repeating the push operation in each case.